Mathlib/NumberTheory/RamificationInertia.lean

```diff
@@ -0,0 +1,852 @@
+  letI : NeZero e := ⟨he⟩
+  have := rank_pow_quot f p P hP0 0 (Nat.zero_le e)
+  rw [pow_zero, Nat.sub_zero, Ideal.one_eq_top, Ideal.map_top] at this
+  exact (rank_top (R ⧸ p) _).symm.trans this
+#align ideal.rank_prime_pow_ramification_idx Ideal.rank_prime_pow_ramificationIdx
+
+/-- If `p` is a maximal ideal of `R`, `S` extends `R` and `P^e` lies over `p`,
+then the dimension `[S/(P^e) : R/p]`, as a natural number, is equal to `e * [S/P : R/p]`. -/
+theorem finrank_prime_pow_ramificationIdx [IsDomain S] [IsDedekindDomain S] (hP0 : P ≠ ⊥)
+    [p.IsMaximal] [P.IsPrime] (he : e ≠ 0) :
+    finrank (R ⧸ p) (S ⧸ P ^ e) =
+      e *
+        @finrank (R ⧸ p) (S ⧸ P) _ _
+          (@Algebra.toModule _ _ _ _ <|
+            @Quotient.algebraQuotientOfRamificationIdxNeZero _ _ _ _ _ _ _ ⟨he⟩) := by
+  letI : NeZero e := ⟨he⟩
+  letI : Algebra (R ⧸ p) (S ⧸ P) := Quotient.algebraQuotientOfRamificationIdxNeZero f p P
+  letI := Ideal.Quotient.field p
+  have hdim := rank_prime_pow_ramificationIdx _ _ _ hP0 he
+  by_cases hP : FiniteDimensional (R ⧸ p) (S ⧸ P)
+  · haveI := hP
+    haveI := (finiteDimensional_iff_of_rank_eq_nsmul he hdim).mpr hP
+    refine' Cardinal.natCast_injective _
+    rw [finrank_eq_rank', Nat.cast_mul, finrank_eq_rank', hdim, nsmul_eq_mul]
+  have hPe := mt (finiteDimensional_iff_of_rank_eq_nsmul he hdim).mp hP
+  simp only [finrank_of_infinite_dimensional hP, finrank_of_infinite_dimensional hPe,
+    MulZeroClass.mul_zero]
+#align ideal.finrank_prime_pow_ramification_idx Ideal.finrank_prime_pow_ramificationIdx
+
+end FactLeComap
+
+section FactorsMap
+
+open scoped Classical
+
+/-! ## Properties of the factors of `p.map (algebraMap R S)` -/
+
+
+variable [IsDomain S] [IsDedekindDomain S] [Algebra R S]
+
+theorem Factors.ne_bot (P : (factors (map (algebraMap R S) p)).toFinset) : (P : Ideal S) ≠ ⊥ :=
+  (prime_of_factor _ (Multiset.mem_toFinset.mp P.2)).ne_zero
+#align ideal.factors.ne_bot Ideal.Factors.ne_bot
+
+instance Factors.isPrime (P : (factors (map (algebraMap R S) p)).toFinset) :
+    IsPrime (P : Ideal S) :=
+  Ideal.isPrime_of_prime (prime_of_factor _ (Multiset.mem_toFinset.mp P.2))
+#align ideal.factors.is_prime Ideal.Factors.isPrime
+
+theorem Factors.ramificationIdx_ne_zero (P : (factors (map (algebraMap R S) p)).toFinset) :
+    ramificationIdx (algebraMap R S) p P ≠ 0 :=
+  IsDedekindDomain.ramificationIdx_ne_zero (ne_zero_of_mem_factors (Multiset.mem_toFinset.mp P.2))
+    (Factors.isPrime p P) (Ideal.le_of_dvd (dvd_of_mem_factors (Multiset.mem_toFinset.mp P.2)))
+#align ideal.factors.ramification_idx_ne_zero Ideal.Factors.ramificationIdx_ne_zero
+
+instance Factors.fact_ramificationIdx_neZero (P : (factors (map (algebraMap R S) p)).toFinset) :
+    NeZero (ramificationIdx (algebraMap R S) p P) :=
+  ⟨Factors.ramificationIdx_ne_zero p P⟩
+#align ideal.factors.fact_ramification_idx_ne_zero Ideal.Factors.fact_ramificationIdx_neZero
+
+set_option synthInstance.checkSynthOrder false
```

Comment:
```suggestion
set_option synthInstance.checkSynthOrder false in
```

Reply:
This does seem to be doing what we think it is doing: adding it make the instance invisible later in the file... For example:
```lean
set_option synthInstance.checkSynthOrder false in -- Porting note: this is okay by the remark below
-- In this file, the value for `f` can be inferred.
attribute [local instance] Ideal.Quotient.algebraQuotientOfRamificationIdxNeZero

@[simp]
theorem Quotient.algebraMap_quotient_of_ramificationIdx_neZero (x : R) :
    algebraMap (R ⧸ p) (S ⧸ P) (Ideal.Quotient.mk p x) = Ideal.Quotient.mk _ (f x) := rfl
```
The `Algebra` structure cannot be synthesize in the theorem. 

I'll leave it the way it is.